File: android/src/main/kotlin/com/djgeo/majascan/g_scanner/ScanFragment.java
package com.djgeo.majascan.g_scanner;

import android.content.DialogInterface;
import android.content.pm.PackageManager;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.widget.Toolbar;
import androidx.core.app.ActivityCompat;
import androidx.fragment.app.Fragment;

import com.djgeo.majascan.R;

import static android.Manifest.permission.CAMERA;
import static com.djgeo.majascan.g_scanner.PermissionUtil.Permission_denied_forever;
import static com.djgeo.majascan.g_scanner.PermissionUtil.Permission_granted;
import static com.djgeo.majascan.g_scanner.QrCodeScannerActivity.REQUEST_CAMERA;

public class ScanFragment extends Fragment implements ScanInteractorImpl.ScanCallbackInterface {

    private String mWebTitle;//如果title不為空 => 顯示webView

    private FrameLayout mCapturePreview;
    private View mScannerBar;
    private CheckBox mFlashlightBtn;
    private TextView mTvTitle;
    private Toolbar mToolbar;
    private ImageView mBackBtn;

    private ScanInteractor scanInteractor;
    private AlertDialog mGoToWebviewDialog;


    public static ScanFragment newInstance(String title, boolean hasFlashLight, int toolBarColor, int titleColor) {

        Bundle args = new Bundle();
        args.putString(QrCodeScannerActivity.TITLE, title);
        args.putBoolean(QrCodeScannerActivity.FLASHLIGHT, hasFlashLight);

        if (toolBarColor != 0) {
            args.putInt(QrCodeScannerActivity.BAR_COLOR, toolBarColor);
        }

        if (titleColor != 0) {
            args.putInt(QrCodeScannerActivity.TITLE_COLOR, titleColor);
        }


        ScanFragment fragment = new ScanFragment();
        fragment.setArguments(args);
        return fragment;

    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_scan, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        scanInteractor = new ScanInteractorImpl(this);
        mCapturePreview = view.findViewById(R.id.capture_preview);
        mScannerBar = view.findViewById(R.id.scan_bar);
        mTvTitle = view.findViewById(R.id.tv_title);
        mToolbar = view.findViewById(R.id.actionbar);

        //閃光燈
        mFlashlightBtn = view.findViewById(R.id.toggle_flashlight);
        mFlashlightBtn.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (isChecked) {
                    scanInteractor.openFlash();
                } else {
                    scanInteractor.closeFlash();
                }

            }
        });

        //關閉按鈕
        mBackBtn = view.findViewById(R.id.back_btn);
        mBackBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (getActivity() != null) {
                    getActivity().finish();
                }
            }
        });

        handleBundleData();
    }

    private void handleBundleData() {
        Bundle args = getArguments();
        if (args != null) {
//            mWebTitle = args.getString(BUNDLE_WEBVIEW_TITLE, "");
            String title = args.getString(QrCodeScannerActivity.TITLE, getString(R.string.scanner_title));
            mTvTitle.setText(title);

            boolean hasFlashLight = args.getBoolean(QrCodeScannerActivity.FLASHLIGHT, true);
            mFlashlightBtn.setVisibility(hasFlashLight ? View.VISIBLE : View.GONE);

            mToolbar.setBackgroundColor(args.getInt(QrCodeScannerActivity.BAR_COLOR, android.R.color.transparent));

            int titleColor = args.getInt(QrCodeScannerActivity.TITLE_COLOR, 0);
            if (titleColor != 0) {
                Drawable drawable = getResources().getDrawable(R.drawable.left_arrow);
                drawable.setColorFilter(new PorterDuffColorFilter(titleColor, PorterDuff.Mode.MULTIPLY));
                mBackBtn.setImageDrawable(drawable);

                mTvTitle.setTextColor(titleColor);
            }

        }
    }

    private void requestPermission() {
        if (getActivity() != null) {
            ActivityCompat.requestPermissions(getActivity(), new String[]{CAMERA}, REQUEST_CAMERA);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[],
                                           int[] grantResults) {
        if (requestCode == REQUEST_CAMERA) {
            if (grantResults.length > 0) {
                boolean cameraAccepted = grantResults[0] == PackageManager.PERMISSION_GRANTED;
                if (cameraAccepted) {
                    startScan();
                }
//                    if (!cameraAccepted) {
//                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
//                            if (shouldShowRequestPermissionRationale(CAMERA)) {
//                                requestPermissions(new String[]{CAMERA}, REQUEST_CAMERA);
//                                return;
//                            }
//                        }
//                    }
            }
        }
    }

    private void showGoToSettingDialog() {
        if (getActivity() != null) {
            new AlertDialog.Builder(getActivity())
                    .setMessage(getString(R.string.camera_permission_tips))
                    .setPositiveButton(getString(R.string.dialog_btn_go), new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                            PermissionUtil.goToSettingPermission(getActivity());
                        }
                    })
                    .setNegativeButton(getString(R.string.dialog_btn_cancel), new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int which) {
                            getActivity().finish();
                        }
                    })
                    .create()
                    .show();
        }

    }

    @Override
    public void onStart() {
        super.onStart();
        int permissionStatus = PermissionUtil.getPermissionStatus(getActivity(), CAMERA);
        switch (permissionStatus) {
            case Permission_granted:
                startScan();
                break;
            default:
                requestPermission();
                break;
            case Permission_denied_forever:
                showGoToSettingDialog();
                break;
        }
    }

    @Override
    public void onStop() {
        super.onStop();
        scanInteractor.stopPreview();
        mScannerBar.clearAnimation();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        scanInteractor.stopPreview();
        if (mGoToWebviewDialog != null) {
            mGoToWebviewDialog.dismiss();
        }
    }

    private void startScan() {
        if (getActivity() == null) return;
        mCapturePreview.removeAllViews();
        scanInteractor.initScan(mCapturePreview);
        scanInteractor.startPreview();

        Animation animation = AnimationUtils.loadAnimation(getActivity(), R.anim.scan_anim);
        mScannerBar.startAnimation(animation);
        animation.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {
                if (mScannerBar != null) {
                    mScannerBar.setVisibility(View.VISIBLE);
                }
            }

            @Override
            public void onAnimationEnd(Animation animation) {
                if (mScannerBar != null) {
                    mScannerBar.setVisibility(View.GONE);
                }
            }

            @Override
            public void onAnimationRepeat(Animation animation) {
            }
        });

    }

    @Override
    public void receiveResult(final String result) {
        final QrCodeScannerActivity activity = (QrCodeScannerActivity) getActivity();
        if (activity == null) return;
        if (!TextUtils.isEmpty(mWebTitle)) {
            //a.方式
            if (mGoToWebviewDialog == null || !mGoToWebviewDialog.isShowing()) {
                AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
                builder.setTitle(R.string.go_to_webview)
                        .setPositiveButton(getString(R.string.dialog_btn_go), new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {
                                activity.goToWebviewFragment(result, mWebTitle);
                            }
                        })
                        .setNegativeButton(getString(R.string.dialog_btn_cancel), null);
                mGoToWebviewDialog = builder.create();
                mGoToWebviewDialog.show();
            }

        } else {
            //b.方式
            activity.receiveAndSetResult(result);
        }
    }
}

File: android/src/main/kotlin/com/djgeo/majascan/g_scanner/ScanFragment.kt
package com.djgeo.majascan.g_scanner


import android.Manifest.permission.CAMERA
import android.app.Activity
import android.content.pm.PackageManager
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.PorterDuffColorFilter
import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.Toolbar
import androidx.core.app.ActivityCompat
import androidx.fragment.app.Fragment
import com.djgeo.majascan.R
import com.djgeo.majascan.g_scanner.QrCodeScannerActivity.Companion.REQUEST_CAMERA
import android.graphics.drawable.GradientDrawable
import android.widget.*
import androidx.core.content.ContextCompat
import kotlin.math.min


class ScanFragment : Fragment(), ScanInteractorImpl.ScanCallbackInterface {

    private val mWebTitle: String? = null

    private var mCapturePreview: FrameLayout? = null
    private var mScannerBar: View? = null
    private var mFlashlightBtn: CheckBox? = null
    private var mTvTitle: TextView? = null
    private var mToolbar: Toolbar? = null
    private var mBackBtn: ImageView? = null
    private var mQrView: QrBorderView? = null
    private var mScanBar: View? = null

    private var scanInteractor: ScanInteractor? = null
    private var mGoToWebviewDialog: AlertDialog? = null

    companion object {

        fun newInstance(title: String, hasFlashLight: Boolean, toolBarColor: Int, titleColor: Int,
                        qrCornerColor: Int = 0, qrScanColor: Int = 0, scanAreaScale: Float): ScanFragment {

            val args = Bundle()
            args.putString(QrCodeScannerActivity.TITLE, title)
            args.putBoolean(QrCodeScannerActivity.FLASHLIGHT, hasFlashLight)

            if (toolBarColor != 0) {
                args.putInt(QrCodeScannerActivity.BAR_COLOR, toolBarColor)
            }

            if (titleColor != 0) {
                args.putInt(QrCodeScannerActivity.TITLE_COLOR, titleColor)
            }

            if (qrCornerColor != 0) {
                args.putInt(QrCodeScannerActivity.QR_CORNER_COLOR, qrCornerColor)
            }

            if (qrScanColor != 0) {
                args.putInt(QrCodeScannerActivity.QR_SCANNER_COLOR, qrScanColor)
            }

            args.putFloat(QrCodeScannerActivity.SCAN_AREA_SCALE, scanAreaScale)

            val fragment = ScanFragment()
            fragment.arguments = args
            return fragment

        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_scan, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        scanInteractor = ScanInteractorImpl(this)
        mCapturePreview = view.findViewById(R.id.capture_preview)
        mScannerBar = view.findViewById(R.id.scan_bar)
        mTvTitle = view.findViewById(R.id.tv_title)
        mToolbar = view.findViewById(R.id.actionbar)
        mQrView = view.findViewById(R.id.capture_crop_view)
        mScanBar = view.findViewById(R.id.scan_bar)

        //閃光燈
        mFlashlightBtn = view.findViewById(R.id.toggle_flashlight)
        mFlashlightBtn?.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) {
                scanInteractor?.openFlash()
            } else {
                scanInteractor?.closeFlash()
            }
        }

        //關閉按鈕
        mBackBtn = view.findViewById(R.id.back_btn)
        mBackBtn?.setOnClickListener { finishActivityWithoutResult() }

        handleBundleData()
    }

    private fun handleBundleData() {
        val args = arguments
        if (args != null && context != null) {
            val title = args.getString(QrCodeScannerActivity.TITLE, getString(R.string.scanner_title))
            mTvTitle?.text = title

            val hasFlashLight = args.getBoolean(QrCodeScannerActivity.FLASHLIGHT, true)
            mFlashlightBtn?.visibility = if (hasFlashLight) View.VISIBLE else View.GONE

            mToolbar?.setBackgroundColor(args.getInt(QrCodeScannerActivity.BAR_COLOR, android.R.color.transparent))

            val titleColor = args.getInt(QrCodeScannerActivity.TITLE_COLOR, 0)
            if (titleColor != 0) {
                val drawable = ContextCompat.getDrawable(context!!, R.drawable.left_arrow)
                drawable?.let {
                    it.colorFilter = PorterDuffColorFilter(titleColor, PorterDuff.Mode.MULTIPLY)
                    mBackBtn?.setImageDrawable(it)
                }
                mTvTitle?.setTextColor(titleColor)
            }

            val qRCornerColor = args.getInt(QrCodeScannerActivity.QR_CORNER_COLOR, 0)
            if (qRCornerColor != 0) {
                mQrView?.setQRCornerColor(qRCornerColor)
            }

            //gradient
            val qRScannerColor = args.getInt(QrCodeScannerActivity.QR_SCANNER_COLOR, Color.rgb(255, 136, 0))
            val gd = GradientDrawable(GradientDrawable.Orientation.LEFT_RIGHT,
                    intArrayOf(qRScannerColor, Color.WHITE, qRScannerColor))
            mScanBar?.background = gd

            //依據BundleData調整mQrView寬高
            var scale = args.getFloat(QrCodeScannerActivity.SCAN_AREA_SCALE)
            if (scale <= 0F) scale = 0F
            if (scale >= 1F) scale = 1F
            val length = (min(resources.displayMetrics.widthPixels,
                    resources.displayMetrics.heightPixels) * scale).toInt()
            val layoutParams = RelativeLayout.LayoutParams(length, length)
            layoutParams.addRule(RelativeLayout.CENTER_IN_PARENT)
            mQrView?.layoutParams = layoutParams
        }
    }

    private fun requestPermission() {
        activity?.let {
            ActivityCompat.requestPermissions(it, arrayOf(CAMERA), REQUEST_CAMERA)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>,
                                            grantResults: IntArray) {
        if (requestCode == REQUEST_CAMERA) {
            if (grantResults.isNotEmpty()) {
                val cameraAccepted = grantResults[0] == PackageManager.PERMISSION_GRANTED
                if (cameraAccepted) {
                    startScan()
                } else {
                    finishActivityWithoutResult()
                }
            }
        }
    }

    private fun showGoToSettingDialog() {
        activity?.let {
            AlertDialog.Builder(it)
                    .setMessage(getString(R.string.camera_permission_tips))
                    .setPositiveButton(getString(R.string.dialog_btn_go)) { _, _ -> activity?.let { PermissionUtil.goToSettingPermission(activity!!) } }
                    .setNegativeButton(getString(R.string.dialog_btn_cancel)) { _, _ -> finishActivityWithoutResult() }
                    .setCancelable(false)
                    .create()
                    .show()
        }
    }

    private fun finishActivityWithoutResult() {
        val activity = activity as QrCodeScannerActivity? ?: return
        activity?.receiveAndSetResult("")
    }

    override fun onStart() {
        super.onStart()
        when (PermissionUtil.getPermissionStatus(activity as Activity, CAMERA)) {
            PermissionUtil.Permission_granted -> startScan()
            PermissionUtil.Permission_denied -> requestPermission()
            PermissionUtil.Permission_denied_forever -> showGoToSettingDialog()
        }
    }

    override fun onStop() {
        super.onStop()
        scanInteractor?.stopPreview()
        mScannerBar?.clearAnimation()
    }

    override fun onDestroy() {
        super.onDestroy()
        scanInteractor?.stopPreview()
        mGoToWebviewDialog?.dismiss()
    }

    private fun startScan() {
        activity?.let {
            mCapturePreview?.removeAllViews()
            scanInteractor?.initScan(mCapturePreview!!)
            scanInteractor?.startPreview()

            val animation = AnimationUtils.loadAnimation(it, R.anim.scan_anim)
            mScannerBar?.startAnimation(animation)
            animation.setAnimationListener(object : Animation.AnimationListener {
                override fun onAnimationStart(animation: Animation) {
                    mScannerBar?.visibility = View.VISIBLE
                }

                override fun onAnimationEnd(animation: Animation) {
                    mScannerBar?.visibility = View.GONE
                }

                override fun onAnimationRepeat(animation: Animation) {}
            })
        }
    }

    override fun receiveResult(result: String) {
        val activity = activity as QrCodeScannerActivity? ?: return
        if (!TextUtils.isEmpty(mWebTitle)) {
            //a.方式
            if (mGoToWebviewDialog == null || !mGoToWebviewDialog!!.isShowing) {
                val builder = AlertDialog.Builder(getActivity()!!)
                builder.setTitle(R.string.go_to_webview)
                        .setPositiveButton(getString(R.string.dialog_btn_go)) { _, _ -> activity.goToWebviewFragment(result, mWebTitle!!) }
                        .setNegativeButton(getString(R.string.dialog_btn_cancel), null)
                mGoToWebviewDialog = builder.create()
                mGoToWebviewDialog!!.show()
            }

        } else {
            //b.方式
            activity.receiveAndSetResult(result)
        }
    }
}
